### ISSUE TITLE
[Bug] DistributedCodegen emits undefined-name Python for tensor.assemble in a Level.HOST orchestrator

### ISSUE THREAD

**Issue by Hzfengsy** [hw-native-sys/pypto#2304]
## Component

Codegen

## Description

A `pl.assemble` written in a `@pl.function(level=pl.Level.HOST, role=pl.Role.Orchestrator)` body reaches `DistributedCodegen`, which has **no `tensor.assemble` handler** registered in `DistributedOpRegistry`. `DistributedCodegen::VisitExpr_(CallPtr)` therefore falls through to its generic call fallback:

```cpp
// src/codegen/distributed/distributed_codegen.cpp:977-978
// Regular op call
current_expr_value_ = op->op_->name_ + "(" + FormatArgs(op->args_) + ")";
```

That line is not a code-emission path — it prints the op as if it were a Python function call. Three of its assumptions are false for `tensor.assemble`:

| Assumption | Reality | Mechanism |
| --- | --- | --- |
| The op name is a Python callable in scope | `tensor` is never imported or bound in the emitted module | `op_->name_` is emitted verbatim |
| Args are plain names/scalars | Tensors live in the `tensors` dict; the offset is a `MakeTuple` | `FormatArgs` (`distributed_codegen.cpp:1499`) and `VisitExpr_(VarPtr)` (`:1002`) emit a bare `SanitizeName(name_hint_)` |
| Kwargs do not matter | `atomic=AtomicType.Add` changes semantics | the fallback formats `args_` only; `kwargs_` is never read |

The result is a single emitted line carrying four undefined names and a silently dropped combine mode.

**The offset is not "flattened" — it is reduced to its last element.** `current_expr_value_` is a single-slot member, and the base `IRVisitor::VisitExpr_(MakeTuplePtr)` (`src/ir/transforms/visitor.cpp:146-151`) visits children in order without composing a result, so the last child overwrites the rest. With a rank-2 offset `[3, 7]` the emission is `..., 7)` — the leading axis vanishes with no diagnostic.

**This is not specific to `assemble`.** Any tensor op absent from `DistributedOpRegistry` hits the same fallback. `pl.reshape(tmp, [8, 32])` in the same body emits `r2__ssa_v0 = tensor.reshape(tmp__ssa_v0, 32)`.

### Why nothing upstream catches it

A HOST-level orchestrator gets `func_type == FunctionType.Opaque` (`python/pypto/language/parser/decorator.py:310-351` — `type=` unset yields `Opaque`), which falls outside every relevant gate:

| Layer | Location | Why it misses |
| --- | --- | --- |
| `ConvertTensorToTileOps` | `convert_tensor_to_tile_ops_pass.cpp:2062` | Phase 1 rewrites `tensor.assemble` only for `FunctionType::InCore`; later phases only patch call sites and param directions |
| `IncoreTileOps` verifier | `convert_tensor_to_tile_ops_pass.cpp:2272` | The check that would flag "tensor op should have been converted" also skips non-InCore functions |
| `FuseCreateAssembleToSlice` | `fuse_create_assemble_to_slice_pass.cpp:377` | Returns the function untouched for non-`Orchestration` |
| `OrchestrationCodegen::HandleTensorAssembleAssign` | `orchestration_codegen.cpp:3806-3828` | Never reached — `python/pypto/backend/pto_backend.py:1316-1332` routes the HOST orchestrator to `DistributedCodegen` and sends only `FunctionType.Orchestration` functions to `_generate_single_chip`. This is the function carrying the atomic guard added for #2256. |
| `VerifyDistributedCodegenPreconditions` | `codegen_preconditions.cpp:126-145` | Checks only comm-domain materialization; there is no op-coverage check |

The structural cause underneath the missing handler: **`DistributedCodegen` has no "unhandled op" failure mode.** Every other backend either lowers an op or rejects it; this one silently degrades to text.

Related: #2256
Related: #2180

## Steps to Reproduce

1. Save the following as `repro.py`:

```python
import pypto.language as pl
import pypto.language.distributed as pld
from pypto.pypto_core import codegen as _codegen_core
from pypto.ir.pass_manager import PassManager, OptimizationStrategy


@pl.program
class P:
    @pl.function(type=pl.FunctionType.Orchestration)
    def chip_orch(self, data: pld.DistributedTensor[[256], pl.FP32]):
        return data

    @pl.function(level=pl.Level.HOST, role=pl.Role.Orchestrator)
    def host_orch(self, out: pl.Tensor[[256], pl.FP32]):
        buf = pld.alloc_window_buffer(256 * pl.FP32.get_byte())
        data = pld.window(buf, [256], dtype=pl.FP32)
        for r in pl.range(pld.world_size()):
            self.chip_orch(data, device=r)
        tmp = pl.create_tensor([256], dtype=pl.FP32)
        out = pl.assemble(out, tmp, [0], atomic=pl.AtomicType.Add)
        return out


for f in P.functions.values():
    print(f"{f.name}: func_type={f.func_type} level={f.level} role={f.role}")

post = PassManager.get_strategy(OptimizationStrategy.Default).run_passes(P)
print(_codegen_core.DistributedCodegen().generate(post))
```

2. Run it:

```bash
PYTHONPATH=$(pwd)/python python repro.py
```

3. Inspect the emitted `host_orch` body.

Two variants confirm the scope of the defect:

- Replace the offset with a multi-axis one — `out: pl.Tensor[[8, 32], pl.FP32]`, `tmp = pl.create_tensor([4, 16], ...)`, `pl.assemble(out, tmp, [3, 7])` — to see the leading offset axis dropped.
- Replace the assemble with `r2 = pl.reshape(tmp, [8, 32])` to see that any unregistered tensor op fails the same way.

## Expected Behavior

A `tensor.assemble` in a HOST orchestrator body should either

- be lowered to correct Python — the `tensor.slice` handler at `src/codegen/distributed/distributed_ops_codegen.cpp:381-438` already establishes the shape for this: walk the `MakeTuple` per element, emit `tensors["lhs"] = tensors["in"][...]`, call `MarkDeclared`. An assemble would be `tensors["out"][o:o+n] = tensors["tmp"]`, with `+=` for `atomic=Add`; or
- be rejected at compile time with an actionable error naming the op and the function.

Either way, compilation must not silently produce a module that raises `NameError` at run time, and an `atomic=` combine mode must never be dropped without a diagnostic.

Independently: the generic fallback at `distributed_codegen.cpp:978` should stop being reachable for `tensor.*` / `tile.*` ops. Without that, the next unregistered op reintroduces the same failure silently.

## Actual Behavior

`ir.compile` succeeds. The generated `host_orch.py` contains:

```python
def _alloc_intermediates(tensors):
    tensors["tmp__ssa_v0"] = torch.zeros((256,), dtype=torch.float32).share_memory_()

def host_orch(orch, _args, config, *, tensors, callables, sub_ids, _keep, world_size, _domain_provider=None):
    ...
        out__ssa_v1 = tensor.assemble(out__ssa_v0, tmp__ssa_v0, 0)
```

Four defects on that one line:

1. `tensor` is not a defined name in the emitted module → `NameError` at run time.
2. `out__ssa_v0` is unbound — `out` is not in the emitted signature at all.
3. `tmp__ssa_v0` is unbound — it is bound as `tensors["tmp__ssa_v0"]` by `_alloc_intermediates`, not as a bare name.
4. `atomic=pl.AtomicType.Add` is gone entirely, with no diagnostic.

Rank-2 offset variant (`pl.assemble(out, tmp, [3, 7])`) — the leading axis is silently discarded:

```python
out__ssa_v1 = tensor.assemble(out__ssa_v0, tmp__ssa_v0, 7)
```

Unregistered-op variant (`pl.reshape(tmp, [8, 32])`) — same failure mode:

```python
r2__ssa_v0 = tensor.reshape(tmp__ssa_v0, 32)
```

## Git Commit ID

6506c89f96d6465ebf574070da3b1aa63d072f1b

## NPU Kind

N/A (not hardware-specific)

## Host Platform

Linux (aarch64)

## Additional Context

Found while reviewing the fix for #2256, which guarded the *Orchestration*-level atomic assemble in `OrchestrationCodegen::HandleTensorAssembleAssign`. This report covers the *HOST*-level path, where `func_type == Opaque` means that guard is never reached — and where the plain (non-atomic) assemble is mangled too, so it is a strictly wider gap than #2256.

#2180 shares the symptom (undefined name in the generated `host_orch.py`) but has a different root cause there — a cross-branch phi of a tensor ref, not an unhandled-op fallback.

Reproduced on a clean build of `6506c89f` (`cmake -B build -DCMAKE_BUILD_TYPE=RelWithDebInfo`). The failure is in host-level Python emission and involves no NPU, so it reproduces without hardware or a simulator.

Filed by a non-owner of the distributed-tensor area; the choice between "register a host-level handler" and "reject with an actionable error" is left to whoever owns that path.

**Comment by lyfne123** [comment]
Fixed by #2409 (merged 2026-08-18, `b20a6980`).

- `tensor.assemble` now has a registered HOST-orchestrator lowering in `DistributedCodegen` that emits tensor-dictionary slice writes, preserving every offset dimension instead of collapsing to the last element.
- The generic call fallback no longer silently degrades to text: any unregistered `tensor.*` / `tile.*` op in `DistributedCodegen` now fails with an actionable diagnostic, which also covers the `pl.reshape` case reported here.
- `atomic=pl.AtomicType.Add` in a HOST orchestrator is explicitly rejected with a message pointing at the InCore scope, rather than having the combine mode dropped.
